main.cpp

```diff
@@ -220,303 +298,351 @@ struct InvoiceManager
     std::string author = "Toby";
 
-    // 1) save a preset
-    bool savePreset(std::string name, std::string type, std::string author); // returns successful save
-    // 2) checks for an existing preset
-    bool checkPreset(std::string name); // returns whether preset exists
-    // 3) rename a preset
-    void renamePreset(std::string oldName, std::string newName);
+    bool savePreset(std::string name, std::string type, std::string author);
+    bool checkPreset(Preset presetA);
+    void renamePreset(std::string newName);
 };
 
-/*
-Thing 8) CPU
-5 properties:
-    1) speed in MHz (float)
-    2) number of cores (int)
-    3) memory in GB (int)
-    4) model (std::string)
-    5) mirocontroller (std::string)
-3 things it can do:
-    1) get drone speed
-    2) balance drone
-    3) execute presets
- */
+bool Preset::savePreset(std::string newName, std::string newType, std::string newAuthor)
+{
+    Preset newPreset;
+    newPreset.name = newName;
+    newPreset.type = newType;
+    newPreset.author = newAuthor;
+
+    return true;
+}
+
+bool Preset::checkPreset(Preset presetA)
+{
+    if ( presetA.size > 0)
+    {
+        return true;
+    }
+
+    return false;
+}
+
+void Preset::renamePreset(std::string newName)
+{
+    if( newName.length() > 0)
+    {
+        name = newName;
+    }
+
+}
 
 struct CPU
 {
-    // 1) speed in MHz (float)
     float speed = 2.1f;
-    // 2) number of cores (int)
     int numCores = 2;
-    // 3) memory in GB (int)
     int memory = 512;
-    // 4) model (std::string)
     std::string model = "U11X";
-    // 5) architecture (std::string)
     std::string architecture = "ARM"; 
 
 
-    // 1) get drone speed
-    float getSpeed(); // returns speed
-    // 2) balance drone
+    float getSpeed();
     void balance();
-    // 3) execute presets
     void runPreset(Preset preset);
 };
 
-/*
-Thing 10) Drone
-5 properties: 
-    1) Propellers
-    2) Battery
-    3) GPS
-    4) CPU
-    5) Presets
-3 things it can do:
-    1) Lift off
-    2) Maneuver 
-    3) Land
- */
+float CPU::getSpeed()
+{
+    float droneSpeed = 0.0f;
+    // connect to GPS and compare position at different times
+    return droneSpeed;
+}
+
+void CPU::balance()
+{
+    // run Propeller.maintain on all relevant propellers
+}
 
+void CPU::runPreset(Preset presetA)
+{
+    std::cout << "Running preset " << presetA.name << ".\n";
+    // execute the preset
+}
 
 struct Drone
 {
-    // 1) Propellers
     Propeller propellerA;
-    // 2) Battery
+    Propeller propellerB;
+    Propeller propellerC;
+    Propeller propellerD;
     Battery batteryA;
-    // 3) GPS
     GPS gpsA;
-    // 4) CPU
     CPU cpuA;
-    // 5) Presets
     Preset presetA;
+    Preset presetLand;
 
-    // 1) Lift off
     void liftOff();
-    // 2) Maneuver
     void maneuver(); 
-    // 3) Land
-    bool land(); // returns if landing area not clear
+    bool land();
 };
 
+void Drone::liftOff()
+{
+    if( int(gpsA.altitude) == 0) // hmm
```

Comment:
No casting until casting is introduced later in the course
If this code produces warnings with the C-style cast, then change the variable types so it doesn't produce warnings or change what it's being compared against.
- [ ] revise